Add std::process

aturon opened the pull request:
Per [RFC 579](https://github.com/rust-lang/rfcs/pull/579), this commit
adds a new `std::process` module. This module is largely based on the
existing `std::old_io::process` module, but refactors the API to use
`OsStr` and other new standards set out by IO reform.

The existing module is not yet deprecated, to allow for the new API to
get a bit of testing before a mass migration to it.

aturon commented:
r? @alexcrichton

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

aturon commented:
This is WIP for two reasons:

1. Not yet passing tests on Windows; it's close.

2. It still uses `old_io` in a few places, since I didn't want to duplicate work in the `std::fs` PR.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+#[cfg(windows)] const MustDieSignal: int = 9;
+/// Signal a process to exit immediately, forcibly killing it. Corresponds to
+/// SIGKILL on unix platforms.
+#[cfg(not(windows))] const MustDieSignal: int = libc::SIGKILL as int;
```
I think that in general we've had pretty good success with trying to have as few `#[cfg]` directives as possible, could this logic be pushed into each respective `sys` module?

tbu- replied:
This should probably not be an `int`.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+
+    stdin: Option<ChildStdin>,
+    stdout: Option<ChildStdout>,
+    stderr: Option<ChildStderr>,
```
In the RFC I think these fields were `public` to allow all sorts of `Option` manipulation of with whatever the available "borrow mode" is for `Child`, was there a reason you avoided that?

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+
+/// A handle to a child procesess's stdin
+pub struct ChildStdin {
+    inner: FileDesc
```
As part of the rewrite I've been trying to purge as many file descriptors from the cross-platform layer as possible, do you think it'd be easy enough to use a `HANDLE` on Windows and a file descriptor on Unix?

aturon replied:
I'm confused, this is just using the `FileDesc` type from `sys::fs` -- what did you have in mind?

alexcrichton replied:
Now that #21936 has landed we've also got [`unix::fd::FileDesc`](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/fd.rs) and [`windows::handle::Handle`](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/handle.rs) for representing the raw representation. Right now we use a bunch of `osfhandle` stuff which we shouldn't be using (because we're using file descriptors). I'm not sure of the precise API to read/write from the handles we have for the child on Windows, but this may be able to contain just a wrapper around `sys::fs::File` (which has a Handle/FileDesc appropriately).

In general though this is a pretty minor implementation detail we can tweak at any time (so long as the API is right, which it will be), so it's not too major.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+        }
+    }
+
+    /// Executes the command as a child process, returning a handle to it.
```
Could this and the below functions reference what happens to the stdio handles if uninitialized? May be good to say that this inherits, output captures, and status() also inherits.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+use fmt;
+use old_io::{self, Reader, Writer};
+use io::{Error, ErrorKind};
+use io::Result as IoResult;
```
I've been stylistically using `io::Result` instead of `IoResult`, I don't mind switching but we should probably be consistent at least.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+use env;
+use ffi::AsOsStr;
+use fmt;
+use old_io::{self, Reader, Writer};
```
Shouldn't this be using `Read` and `Write`?

aturon replied:
Not yet, because it's using the old `sys::fs::FileDesc`. (Note: I mentioned this in the top comment on the PR.)

alexcrichton replied:
Gah sorry, I should read more closely, I'll comment below.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+#![unstable(feature = "process", reason = "recently added via RFC 579")]
+#![allow(non_upper_case_globals)]
+
+use core::prelude::*;
```
Could this actually import the real prelude? (it's super convenient when it works!)

I've also been trying to make it a habit of importing the I/O prelude into these modules as well (for read/write/etc)

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(f, "{:?}", self.inner.program));
+        for arg in &self.inner.args {
+            try!(write!(f, " '{:?}'", arg));
```
One thing I've noticed recently is that the surrounding single-ticks are actually duplicating the quotes inserted by `:?`, so I think they can be removed.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    ///
+    /// Never generated on Windows.
+    Signal(i32),
+}
```
The RFC mentioned that this representation would be hidden, was that just forgotten?

aturon replied:
No, I changed my mind on this. In particular, the API the RFC suggested was almost identical to this, except that rather than two variants you just got an `Option` of the first variant. Given that, I don't think there's a good reason to hide these details.

alexcrichton replied:
I find it a little odd that we're exposing `Signal` which is a purely Unix-only thing (e.g. it's not through an extension trait). Taking a look at the man page for `wait` though we may want to hide the representation to allow us to tweak it in the future or expose some of the macros like `WCOREDUMP` or possibly `WSTOPSIG` if we get around to it.

I imagine we may be able to pack this kind of information in somewhere, it just seemed like a good conservative step to take without losing much expressiveness (it's pretty rare anyway to match on this).

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    /// This stream will be ignored. This is the equivalent of attaching the
+    /// stream to `/dev/null`
+    Null,
+}
```
The RFC mentioned that this would have static constructors so we could add more variants in the future, was that forgotten?

aturon replied:
Note the `FIXME`: I would prefer to use extensible enums instead.

alexcrichton replied:
Hm true, but perhaps we should use constructor functions until then? (or are we pretty confident we'll get them soon?)

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+                }
+            }
+        }
+        #[cfg(windows)] fn collect_status(_p: &mut Child) {}
```
Could this be pushed down into each respective `sys` module?

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    /// # Errors
+    ///
+    /// This function can fail if a timeout was previously specified via
+    /// `set_timeout` and the timeout expires before the child exits.
```
These docs about `set_timeout` may be a little dated.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    pub fn wait_with_output(mut self) -> IoResult<Output> {
+        drop(self.stdin.take());
+        fn read<T: Reader + Send>(stream: Option<T>) -> Receiver<old_io::IoResult<Vec<u8>>> {
+            let (tx, rx) = channel();
```
Now that we have `JoinGuard` this may actually be able to eschew the channel entirely. I've been uncomfortable in the past about spawning threads here, but I think we do have a theoretical path forward where we don't need to spawn threads, so it may not be so bad.

alexcrichton commented on src/libstd/sys/common/mod.rs:
```diff
@@ -74,6 +75,15 @@ pub fn mkerr_libc<T: Int>(ret: T) -> IoResult<()> {
+    } else {
+        Ok(())
+    }
+}
```
I [also added this](https://github.com/rust-lang/rust/pull/21936/files#diff-28f2fd684ad47d385427678d96d2dcd4R182) as part of `std::fs`, so we may just want to be sure to coordinate on it.

alexcrichton commented on src/libstd/sys/unix/ext.rs:
```diff
@@ -121,20 +128,60 @@ impl OsStringExt for OsString {
     fn as_byte_slice(&self) -> &[u8];
+
+    /// Convert the `OsStr` slice into a `CString`.
+    fn to_cstring(&self) -> CString;
```
`to_c_string`? I think elsewhere you referred to `into_os_string`

aturon replied:
Hm, but that would be somewhat odd for a slice?

alexcrichton replied:
Oh just in the sense for an underscore between `c` and `string` because for `OsString` it was underscore'd as `os_string`

aturon replied:
See http://aturon.github.io/style/naming/README.html -- we have a special case for single-letter "words".

alexcrichton replied:
Aha1

alexcrichton commented on src/libstd/sys/unix/ext.rs:
```diff
@@ -121,20 +128,60 @@ impl OsStringExt for OsString {
+/// Unix-specific extensions to `OsStr`.
 pub trait OsStrExt {
+    /// Get the underlying byte view of the `OsStr` slice.
     fn as_byte_slice(&self) -> &[u8];
```
I would personally have expected `.as_bytes()` here, although we can also tweak it later in a future stabilization PR.

aturon replied:
Happy to rename.

alexcrichton commented on src/libstd/sys/unix/os.rs:
```diff
@@ -429,6 +430,11 @@ pub unsafe fn pipe() -> IoResult<(FileDesc, FileDesc)> {
 
+// FIXME: remove the error shim once old_io is gone
+pub unsafe fn pipe_shim() -> io::Result<(FileDesc, FileDesc)> {
+    pipe().map_err(|_| io::Error::last_os_error())
```
I think this may not actually work out because `IoError::last_error` may allocate and could change errno.

alexcrichton commented on src/libstd/sys/unix/process2.rs:
```diff
@@ -0,0 +1,412 @@
+
+/// The unique id of the process (this should never be negative).
+pub struct Process {
+    pub pid: pid_t
```
Enjoy loving on the edge, do we?

aturon replied:
Not sure I follow?

alexcrichton replied:
Oh just putting `pub` on all the fields.

alexcrichton commented on src/libstd/sys/unix/process2.rs:
```diff
@@ -0,0 +1,412 @@
+        })
+    }
+
+    pub fn wait(&self, _deadline: u64) -> IoResult<ExitStatus> {
```
Perhaps `_deadline` could just be removed?

aturon replied:
I left in the Windows side of the functionality for now, since I anticipate providing it via an extension trait, but I can try to refactor this API down.

alexcrichton commented on src/libstd/sys/unix/process2.rs:
```diff
@@ -0,0 +1,412 @@
+    pub fn wait(&self, _deadline: u64) -> IoResult<ExitStatus> {
+        let mut status = 0 as c_int;
+        match retry(|| unsafe { c::waitpid(self.pid, &mut status, 0) }) {
+            -1 => panic!("unknown waitpid error: {:?}", super::last_error()),
```
We may want to actually return the error here (I wrote a [helper for this](https://github.com/rust-lang/rust/pull/21936/files#diff-28f2fd684ad47d385427678d96d2dcd4R191) as well)

alexcrichton commented:
I added a few comments about the implementation details themselves, but we can also continue to hammer those out later, I think that the API is the more important part to land as part of this PR.

nagisa commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    pub fn env<S: ?Sized, T: ?Sized>(&mut self, key: &S, val: &T) -> &mut Command where
+        S: AsOsStr, T: AsOsStr
+    {
+        self.inner.init_env_map(env::vars());
```
This smells. This initialisation should be encapsulated inside the `inner` implementation.

nagisa commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+pub struct Child {
+    handle: ProcessImp,
+
+    /// None until wait() is called.
```
It will also become `Some` after `wait_with_output`.

nagisa commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+}
+
+impl Child {
+    /// Sends a signal to this child forcing it to exit. This is equivalent to
```
“Sends a signal” exposes a certain implementation detail which should not be exposed. Especially given there’s [TerminateProcess](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686714%28v=vs.85%29.aspx).

nagisa commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    /// output on the stdout/stderr handles, returning a `Output`
+    /// instance.
+    ///
+    /// The stdin handle to the child is closed before waiting.
```
Why?

nagisa commented on src/libstd/sys/unix/ext.rs:
```diff
@@ -121,20 +128,60 @@ impl OsStringExt for OsString {
+
+pub trait CommandExt {
+    /// Sets the child process's user id. This translates to a `setuid` call in
+    /// the child process. Setting this value on windows will cause the spawn to
```
Why is windows specific behaviour is documented in unix module?

nagisa commented on src/libstd/sys/windows/process2.rs:
```diff
@@ -0,0 +1,553 @@
+
+impl Process {
+    pub unsafe fn kill(&self, signal: int) -> IoResult<()> {
+        Process::killpid(self.pid, signal)
```
This is bad.

huonw replied:
Could you be more specific?

nagisa replied:
`struct Process` already contains the `Handle` and instead of using that handle, this will instead use pid to retrieve the handle from the OS. This introduces an unnecessary failure branch and is possibly racy.

alexcrichton replied:
I think that @aturon was still working on the Windows implementation, but I agree that this method should avoid `OpenProcess` if possible and use the `HANDLE` already contained within.

alexcrichton commented on src/libstd/sys/windows/process2.rs:
```diff
@@ -0,0 +1,553 @@
+        unsafe {
+            let process = OpenProcess(SYNCHRONIZE | PROCESS_QUERY_INFORMATION,
+                                      FALSE,
+                                      self.pid as DWORD);
```
One other implementation detail it may be nice to smooth over from the past is that I think this call to `OpenProcess` should be unnecessary (`self` already stores a `HANDLE`)

vojtechkral commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,835 @@
+    fn test_process_output_fail_to_start() {
+        match Command::new("/no-binary-by-this-name-should-exist").output() {
+            Err(e) => assert_eq!(e.kind(), ErrorKind::FileNotFound),
+            Ok(..) => panic!()
```
Out of curiosity, is #21571 addressed here?

aturon commented:
@alexcrichton I believe I've addressed the majority of your comments, modulo a couple things we discussed today.

@nagisa Note that most of the code here is inherited from the existing implementation, not written from scratch. I've put in a round of cleanup, but there are a number of places in the implementation that could be improved. Notably, we should try to use named pipes on windows directly to avoid use of file descriptors here. But the priority at the moment is landing the surface-level APIs; we can improve these implementation details further later on.

That said, I believe I've addressed the concerns you raised.

aturon commented:
Also, this is now using the new IO traits, etc, since it's rebased over `std::fs`. It passes tests on mac os, and compiles on Windows; I should know whether it passes Windows tests shortly.

aturon commented:
@alexcrichton Note, this is failing tests on Windows, so don't send it to bors just yet. (But modulo figuring out the problem, it's otherwise ready for a second review.)

alexcrichton commented on src/libstd/process.rs:
````diff
@@ -0,0 +1,837 @@
+/// to be changed (for example, by adding arguments) prior to spawning:
+///
+/// ```
+/// use std::old_io::Command;
````
Docs may need to be updated.

alexcrichton commented on src/libstd/process.rs:
````diff
@@ -0,0 +1,837 @@
+///
+/// let output = process.stdout.as_mut().unwrap().read_to_end();
+/// ```
+#[derive(Clone)]
````
We may actually want to consider removing this `Clone` implementation. I've thought in the past that if `stdin`, `stdout`, and `stderr` happen to take some form of I/O object it will likely not implement `Clone`.

Although we could also just say that the `Clone` implementation clears the input/output handles.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,837 @@
+    /// Executes the command as a child process, returning a handle to it.
+    ///
+    /// By default, stdin, stdout and stderr are inherited by the parent.
+    pub fn spawn(&self) -> io::Result<Child> {
```
Hm, I forgot about this earlier, but if `Stdio` holds an I/O object and needs to transfer ownership that in theory could affect the `&self` here. We may want to consider taking `&mut self`, but on the other hand the worst case scenario would be adding a `RefCell` which wouldn't necessarily be the end of the world.

alexcrichton commented on src/libstd/process.rs:
```diff
@@ -0,0 +1,837 @@
+}
+
+/// Describes what to do with a standard io stream for a child process.
+#[derive(Clone)]
```
Similar worries about `Clone` here as above with `Command`

alexcrichton commented on src/libstd/sys/unix/ext.rs:
```diff
@@ -130,19 +136,28 @@ impl OsStringExt for OsString {
+    fn as_bytes(&self) -> &[u8];
+
+    /// Convert the `OsStr` slice into a `CString`.
+    fn to_cstring(&self) -> CString;
```
Thinking twice on this, we currently don't provide other convenience constructors for `CString` (only a `from_slice` constructor basically). In some sense this may be better placed on a `ToCString` trait, but you still need a clear signal that this is platform-specific behavior. All-in-all we may want to leave this out for now and perhaps add it back later if necessary.

aturon replied:
Incidentally, this is one reason that generic conversion traits are helpful (you don't have this kind of overlap worry).

But you're also raising an important general point, which is that we can't provide opt-in platform-specific impls for well-known traits.

Personally, I believe that this convenience is very important, because when you're working with `OsStr` on a given platform you need it to be very easy to convert that to something you can feed the OS. (I've already heard complaints about its lack). I'd somewhat prefer to land this now, but revisit before we stabilize.

alexcrichton replied:
Ok, sounds good to me!

alexcrichton commented:
Just a few more minor nits, but otherwise r=me

aturon commented:
@alexcrichton Updated, removing `FileDesc` from Windows and introducing `pipe2` modules for all platforms. At the moment, `pipe2` covers only `AnonPipe` but we can later migrate names pipes/domain sockets.

I also moved `read`/`write` into `Handle` for Windows, and made `read` deal with `BrokenPipe` correctly.

aturon commented:
@alexcrichton I have also addressed all of your comments, and tidied up a few names (the fields of `Output`, mostly).

alexcrichton commented on src/libstd/sys/windows/handle.rs:
```diff
@@ -21,7 +21,37 @@ impl Handle {
     }
+
+    pub unsafe fn from_fd(fd: c_int) -> Handle {
+        libc::get_osfhandle(fd) as libc::HANDLE
```
I think that MSVCRT requires that file descriptors have `close` called on them to free both the backing `HANDLE` as well as MSVCRT resources, so this function may not be able to exist :(

alexcrichton commented on src/libstd/sys/windows/handle.rs:
```diff
@@ -21,7 +21,37 @@ impl Handle {
+
+        match res {
+            Ok(_) => Ok(read as usize),
+            Err(e) if e.kind() == ErrorKind::BrokenPipe => Ok(0),
```
Could you add a comment here explaining why it's handled?

alexcrichton commented:
Just a few last nits, but otherwise looks good to go to me. r=me with nits addressed

tanadeau commented on src/libstd/process.rs:
````diff
@@ -0,0 +1,828 @@
+/// # Example
+///
+/// ```should_fail
+/// use std::io::Command;
````
The example code in the docs should go from `use std::io` to `use std::process`

aturon commented:
@bors: r=alexcrichton cc408a0

bors commented:
:hourglass: Testing commit cc408a0 with merge c748266...

bors commented:
:broken_heart: Test failed - [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3566)

aturon commented:
@bors: r=alexcrichton da46471

bors commented:
:hourglass: Testing commit da46471 with merge 0d440a2...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3575)

aturon commented:
@bors: r=alexcrichton 7defdfa

bors commented:
:hourglass: Testing commit 7defdfa with merge 2978d79...

alexcrichton commented on src/libstd/sys/windows/process2.rs:
```diff
@@ -0,0 +1,487 @@
+pub struct Process {
+    /// A HANDLE to the process, which will prevent the pid being
+    /// re-used until the handle is closed.
+    handle: *mut c_void,
```
If this bounces this may want to change to `sys::handle::Handle` as it appropriately handles `Send`/`Sync` and provides RAII drop as well (cc https://github.com/rust-lang/rust/issues/22228)

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3572)

aturon commented:
@bors: r=alexcrichton  02666fb

bors commented:
:hourglass: Testing commit 02666fb with merge 2687494...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3585)

aturon commented:
@bors: r=alexcrichton 1a72dac

aturon commented:
@bors: p=1 (due to upcoming release)

bors commented:
:hourglass: Testing commit 1a72dac with merge 7b780ce...

bors commented:
:broken_heart: Test failed - [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3243)

bors commented:
:broken_heart: Test failed - [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/2073)

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3248)

bors commented:
:broken_heart: Test failed - [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2081)

aturon commented:
@bors: retry

bors commented:
:hourglass: Testing commit 1a72dac with merge d05755d...

bors commented:
:broken_heart: Test failed - [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3245)

aturon commented:
@bors: r=alexcrichton 0f697c2 p=1

bors commented:
:hourglass: Testing commit 0f697c2 with merge 1ac4113...

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3265)

aturon commented:
@bors: r=alexcrichton bec4af2 p=1

bors commented:
:hourglass: Testing commit bec4af2 with merge 972bafc...

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3573)

aturon commented:
@bors: r=alexcrichton 4175f1c p=1

bors commented:
:hourglass: Testing commit 4175f1c with merge df54632...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/3598), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3602), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/3593), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3601), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3576), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3613), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/3618), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3610), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3270), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3265), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/2095), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2103)
